cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNode.java

```diff
@@ -16,16 +16,24 @@
 
 package co.cask.cdap.api.workflow;
 
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import javax.annotation.Nullable;
+
 /**
  * Represents node in the {@link Workflow}.
  */
 public abstract class WorkflowNode {
   protected final String nodeId;
   protected final WorkflowNodeType nodeType;
+  protected final Set<String> parentNodeIds;
 
-  public WorkflowNode(String nodeId, WorkflowNodeType nodeType) {
+  public WorkflowNode(String nodeId, WorkflowNodeType nodeType, @Nullable Set<String> parentNodeIds) {
```

Comment:
when would this be null? Why not require that it is the empty set if there are no parents? (That would only be the case for the first node in a workflow, right?)